File: C++/TranThong.cc
#include "TranThong.h"
#include <functional>
#include <cstdlib>

static inline int sign(int x) {
  return (x > 0) - (x < 0);
}

void TranThong(
  int x0, int y0,
  int x1, int y1,
  std::function<void(int, int)> callback
) {
  int dx = x1 - x0;
  int dy = y1 - y0;

  int sx = sign(dx);
  int sy = sign(dy);

  dx = std::abs(dx);
  dy = std::abs(dy);

  int x = x0;
  int y = y0;

  callback(x, y);

  if (dx >= dy) {
    // x-major
    int err = dx - dy;
    for (int i = 0; i < dx; ++i) {
      x += sx;
      err -= 2 * dy;
      if (err < 0) {
        y += sy;
        err += 2 * dx;
      }
      callback(x, y);
    }
  } else {
    // y-major
    int err = dy - dx;
    for (int i = 0; i < dy; ++i) {
      y += sy;
      err -= 2 * dx;
      if (err < 0) {
        x += sx;
        err += 2 * dy;
      }
      callback(x, y);
    }
  }
}

File: Lua/TranThong.lua
--[[
  TranThong algorithm
  (a symmetric version of Bresenham's line algorithm)
  
  Reference:
  Thong, Tran. "A symmetric linear algorithm for line segment generation."
  Computers & Graphics 6.1 (1982): 15-17.
]]

local function abs(x) return x < 0 and -x or x end

local function sign(x)
  if x < 0 then return -1 end
  if x > 0 then return 1 end
  return 0
end

local function TranThong(x0, y0, x1, y1, callback)
  local dx = x1 - x0
  local dy = y1 - y0

  local sx = sign(dx)
  local sy = sign(dy)

  dx = abs(dx)
  dy = abs(dy)

  local x = x0
  local y = y0

  callback(x, y)

  if dx >= dy then
    -- x-major
    local e = dx - dy
    for _ = 1, dx do
      x = x + sx
      e = e - 2 * dy
      if e < 0 then
        y = y + sy
        e = e + 2 * dx
      end
      callback(x, y)
    end
  else
    -- y-major
    local e = dy - dx
    for _ = 1, dy do
      y = y + sy
      e = e - 2 * dx
      if e < 0 then
        x = x + sx
        e = e + 2 * dy
      end
      callback(x, y)
    end
  end
end

return TranThong
